infra/src/test/kotlin/fi/fta/geoviite/infra/ui/util/Browser.kt

```diff
@@ -0,0 +1,173 @@
+import fi.fta.geoviite.infra.ui.pagemodel.common.SCREENSHOTS_PATH
+import io.github.bonigarcia.wdm.WebDriverManager
+import org.apache.commons.io.FileUtils
+import org.json.JSONObject
+import org.openqa.selenium.JavascriptExecutor
+import org.openqa.selenium.OutputType
+import org.openqa.selenium.TakesScreenshot
+import org.openqa.selenium.WebDriver
+import org.openqa.selenium.chrome.ChromeDriver
+import org.openqa.selenium.chrome.ChromeOptions
+import org.openqa.selenium.devtools.v113.emulation.Emulation
+import org.openqa.selenium.firefox.FirefoxDriver
+import org.openqa.selenium.firefox.FirefoxOptions
+import org.openqa.selenium.logging.LogEntries
+import org.openqa.selenium.logging.LogEntry
+import org.openqa.selenium.logging.LogType
+import org.openqa.selenium.logging.LoggingPreferences
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import java.io.File
+import java.time.Instant
+import java.util.concurrent.atomic.AtomicReference
+import java.util.logging.Level
+
+private fun createChromeDriver(headless: Boolean): WebDriver {
+    WebDriverManager.chromedriver().setup()
+    val options = ChromeOptions()
+
+    if (headless) options.addArguments("--headless")
+    //if (!headless) chromeOptions.addArguments("--app=http://localhost:9000")
+    options.addArguments("--disable-dev-shm-usage")
+    options.addArguments("--no-sandbox")
+    options.addArguments("--whitelisted-ips=")
+    options.addArguments("--window-size=2560,1640")
+    options.addArguments("--incognito")
+    options.setExperimentalOption("excludeSwitches", listOf("enable-automation"))
+    options.addArguments("--remote-allow-origins=*")
+
+    //if (!headless) chromeOptions.addArguments("--auto-open-devtools-for-tabs")
+
+    val logPrefs = LoggingPreferences()
+    logPrefs.enable(LogType.BROWSER, Level.ALL)
+    logPrefs.enable(LogType.PERFORMANCE, Level.ALL)
+    options.setCapability("goog:loggingPrefs", logPrefs)
+
+    val driver = ChromeDriver(options)
+    val devTools = driver.devTools
+    devTools.createSession()
+    devTools.send(Emulation.setTimezoneOverride("Europe/Helsinki"))
+
+    return driver
+}
+
+private fun createFirefoxDriver(headless: Boolean): WebDriver {
+    WebDriverManager.firefoxdriver().setup()
+    val firefoxOptions = FirefoxOptions()
+    if (headless) firefoxOptions.addArguments("-headless")
+    firefoxOptions.addArguments("-private")
+    firefoxOptions.addArguments("-width=2560")
+    firefoxOptions.addArguments("-height=1640")
+    firefoxOptions.addArguments("-purgecaches")
+    firefoxOptions.addArguments("-safe-mode")
+    return FirefoxDriver(firefoxOptions)
+}
+
+private val logger: Logger = LoggerFactory.getLogger("BROWSER")
+private val webDriver: AtomicReference<WebDriver> = AtomicReference(null)
+
+private fun setBrowser(createWebDriver: () -> WebDriver?) {
+    webDriver.updateAndGet { currentDriver ->
```

Comment:
kaikki selaimen alustukset menee nyt tämän kautta ja se varmista että samalla tapetaan edellinen

Reply:
Aiemmin tuossa quitin yhteydessä muuten odoteltiin vähä aikaa jostain syystä. Voisin kokeilla vielä auttaisiko se testejä menemään läpi

Reply:
testasin ja ei tuntunut auttavan. tuntuu sulkevan ok ilmankin ja testit failaa jostain muusta johtuen

Reply:
Saisiko tähän (tai jonnekin muualle geneeriseen testienajoketjuun) vielä jonkun option, että selain jäisikin päälle? Screenshotit ovat hyvä lisä varsinkin AWS:ään, mutta jos pitäisi päästä lokaalisti tutkimaan mikä mättää DOM:issa, niin melkein tarvitsisi failure-kohtaan selaimen mistä saa devtoolsin auki

Reply:
Sille on jo joku selenium propsu olemassa. Muistan nähneeni jossain mutta en nyt muista ulkoa mikä se oli. Se sotkee tietty niitä ajoja siitä eteenpäin että pitäisi melkein saada se stoppaamaan siihen failanneeseen. En säädä tässä yhteydessä mutta voisi olla kyllä ihan hyvä tehdä noista joku helppo optio

Reply:
Juu, selain auki pitääkin melkein ajella näitä testejä yksi kerrallaan. Lähinnä mietin lokaalia debuggaamista, sillä ainakin itse olen monesti hyötynyt siitä että pääsee oikeasti räpeltämään DOM:ia siitä kun jotain hajoaa